### PR TITLE
New Reduction Rule: General SAT to 3-SAT

### PR DESCRIPTION
**Pull request by SciCodePhy**

### CODE REVIEW

File: src/rules/rules.jl [hunk 1]
```diff
@@ -48,5 +48,7 @@ function extract_solution end
 
 include("spinglass_sat.jl")
 include("spinglass_maxcut.jl")
+include("vertexcovering_setcovering.jl")
+include("sat_3sat.jl")
 include("sat_coloring.jl")
-include("vertexcovering_setcovering.jl")
\ No newline at end of file
+include("vertexcovering_setcovering.jl")
```
Comment (c-allergic) [inline src/rules/rules.jl:54]: clean up

File: src/ProblemReductions.jl [hunk 2]
```diff
@@ -29,7 +29,9 @@ export QUBO
 # rules
 export target_problem, AbstractProblem, AbstractReductionResult, reduceto, extract_solution, reduction_complexity
 export LogicGadget, truth_table
+
 export ReductionCircuitToSpinGlass, ReductionMaxCutToSpinGlass, ReductionSpinGlassToMaxCut, ReductionVertexCoveringToSetCovering, ReductionSatToColoring
+export ReductionSATTo3SAT, rename_variables, transform_to_3_literal_clause, transform_to_3_literal_cnf, generate_dummy_var
```
Comment (GiggleLiu) [inline src/ProblemReductions.jl:34]: Some functions not exposed to users should not be exported.

File: src/rules/sat_3sat.jl [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+function reduceto(::Type{<:Satisfiability}, sat_source::Satisfiability)
+    sat_source_renamed, new_var_map, inverse_new_var_map = rename_variables(sat_source)
+    sat_target = transform_to_3_literal_cnf(sat_source_renamed)
+    return ReductionSATTo3SAT(sat_source, sat_target, new_var_map, inverse_new_var_map )
+end
+
+function extract_solution(res::ReductionSATTo3SAT, sol)
+     
+    # inverse_map = res.inverse_new_var_map
```
Comment (GiggleLiu) [inline src/rules/sat_3sat.jl:23]: Clean up comments.

File: src/rules/sat_3sat.jl [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+"""
+struct ReductionSATTo3SAT{T}
+    sat_source::Satisfiability{T}
+    sat_target::Satisfiability{T}
+    new_var_map::Dict{Symbol, Symbol}
+    inverse_new_var_map::Dict{Symbol, Symbol}
+end
+target_problem(res::ReductionSATTo3SAT) = res.sat_target
+
+function reduceto(::Type{<:Satisfiability}, sat_source::Satisfiability)
```
Comment (GiggleLiu) [inline src/rules/sat_3sat.jl:15]: Maybe you want to reduce the problem to KSatisfiability problem?

File: src/rules/sat_3sat.jl [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+        push!(all_original_solutions, original_solution)
+    end
+
+    return unique( all_original_solutions )
+end
+
+# ----Useful functions----
+# 001: Function to rename variables in the CNF
+function rename_variables(sat::Satisfiability)
```
Comment (GiggleLiu) [inline src/rules/sat_3sat.jl:55]: Please check the `gensym` function:
https://docs.julialang.org/en/v1/base/base/#Base.gensym

File: test/rules/rules.jl [hunk 6]
```diff
@@ -14,6 +14,10 @@ end
     include("sat_coloring.jl")
 end
 
+@testset "sat_3sat" begin
+    include("sat_3sat.jl")
+end
+
 @testset "rules" begin
```
Comment (GiggleLiu) [inline test/rules/rules.jl:21]: Pleaes also add a test here.

File: test/rules/sat_3sat.jl [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    @test transformed_cnf == expected_transformed_cnf
+
+    reduction_result = reduceto(Satisfiability, sat01)
+    @test reduction_result.sat_target == Satisfiability( transform_to_3_literal_cnf( rename_variables(sat01)[1] ).cnf )
+    @test target_problem( reduction_result ) == Satisfiability( transform_to_3_literal_cnf( rename_variables(sat01)[1] ).cnf )
+
+    original_sol = findbest(sat01, BruteForce() )
+    new_sol = findbest(reduction_result.sat_target, BruteForce() )
+    @test original_sol == extract_solution(reduction_result, new_sol)
+end
```
Comment (GiggleLiu) [inline test/rules/sat_3sat.jl:43]: Well tested, good job!

File: src/ProblemReductions.jl (hunk 2 above)
Comment (SciCodePhy) [reply src/ProblemReductions.jl]: Resolved

File: src/rules/sat_3sat.jl (hunk 3 above)
Comment (SciCodePhy) [reply src/rules/sat_3sat.jl]: Resolved

File: src/rules/sat_3sat.jl (hunk 5 above)
Comment (SciCodePhy) [reply src/rules/sat_3sat.jl:50]: Maybe we can temporarily keep the "rename_variables" function.

File: src/rules/sat_3sat.jl (hunk 4 above)
Comment (SciCodePhy) [reply src/rules/sat_3sat.jl]: Resolved: Separate the reduction into SAT=>3-SAT and k-SAT=>SAT

File: test/rules/rules.jl (hunk 6 above)
Comment (SciCodePhy) [reply test/rules/rules.jl]: Resolved: add two tests here

**Comment by SciCodePhy** [comment]
Achieved:
- Separate the reduction into SAT->3-SAT and k-SAT->SAT;
- Add more interfaces in KSatisfiability;
- Add more tests (both in rules.jl and individual test files);
- Modify the extract_solution function such that it can handle both the [sol] (the set of individual solutions) and an individual solution;

Note:
- The SAT part is not very consistent with the new PR #68 . We can merge this PR (SAT -> 3-SAT) first and update the corresponding part later;
- I want to temporarily keep the "rename_variables" function rather than using "gensym".

**Review by GiggleLiu — APPROVED** [review]
Looks good.

**Pull request merged**